### ISSUE TITLE
GPU OOM error while allocating logits for a batch in NMT

### ISSUE THREAD

**Issue by abhudev** [vineetm/ell-881-2018-deep-learning#33]
In NMT, the TensorFlow is showing out of memory error while trying to allocate logits for a batch. Output (german) vocab size is 30000 so the logits look like [batch_size, sentence_length, 30000]. In my case, size of [32,92,30000] takes up 331 MB. Funny thing is, Usage is only 8.6 GB whereas about 11 GB are still available. However, I am concerned that even with 11 GB memory usage may grow eventually, causing a crash. How do I deal with this error, and reduce GPU memory usage in general?

**Comment by prateek27** [comment]
Faced similar issue, some batches have quite large sentence_length, which gives OOM, try reducing batch size.
If it still doesn't work, skip the batches with sentence_length > some_threshold_length.

**Comment by abhudev** [comment]
OK thank you, is 92 sentence length too long? Also, could you figure out if this error is due to memory leakage, and we can somehow handle it? Batch size is 32 (as mentioned in project desctiption) plus reducing it will give even slower training than is now happening. I will try both the techniques you mentioned.

**Comment by abhudev** [comment]
I saved the model checkpoint after NLI, and restored it before starting again for NMT (new run of program), but it still gives OOM error after a few 100 batches. I have reduced batch size to 20 and max sentence length to 45, still it is giving error. @vineetm Sir I have also tried other tricks as mentioned online such as making variables None before returning from functions, but nothing seems to work to solve the OOM error, could you please give us some guidelines on how to resolve this?

**Comment by vineetm** [comment]
@abhudev What are the sizes of the first two dimensions of OOM failure? One could reduce precision. I think default if `tf.float64`, this could be made to `tf.float32`. That should reduce space requirements by half.

**Comment by abhudev** [comment]
Yes Sir, that is a very good suggestion, I will try it out. The first two sizes were, [32, 92, 30000] and [20, 126, 30000] in two separate cases.

**Comment by vineetm** [comment]
@abhudev So clearly 92 and 126 in second dimensions is a problem. I would recommend to restrict `max_words` to 50 or 45 (whatver can fit in the memory). This can be done easily in your dataset pipelinine after you get words from a sentence by doing something as `words[:MAX_LEN]` inside map function. See Line 151 here for an example: https://github.com/tensorflow/nmt/blob/master/nmt/utils/iterator_utils.py

**Comment by abhudev** [comment]
Thank you very much Sir!
